File: nn/nn.py
"""
Caroline Danzi
Dr. Zmuda
CSE 465
Nearest Neighbor Problem
"""

import sys
import re
import math

class DataError(Exception):
    def __init__(self, msg):
        self.msg = msg
    def __str__(self):
        return msg

# Input: The file containing the prototype descriptions
# Output: A 2D array of the prototypes - each prototype
# is represented as an array of its data with the element
# at the last index containing the category
def process_prototypes(proto_file):
    # Loop through each line in prototype file
    prototype_file = open(proto_file, 'r')
    prototypes = []
    
    for line in prototype_file:
        line = re.sub('\s*', '', line)
        proto = line.split(',')
        prototypes.append(proto)

    return prototypes

# Takes in the list of prototypes and the file containing
# the unknowns and prints out the category for each
# unknown based on the prototype it is closest to
def find_nearest_neighbor(prototypes, unknown_file):
    unknowns = open(unknown_file, 'r')
    for line in unknowns:
        line = re.sub('\s*', '', line)
        unknown = line.split(',')
        category = categorize(unknown, prototypes)
        print(category)

# Input: An unknown (as a string list), the list of prototypes
# Output: The category of the prototype to which the unknown
# is closest
def categorize(unknown, prototypes):
    min_dist = sys.maxsize
    category = None
    for prototype in prototypes:
        dist = compute_dist(prototype, unknown)
        if dist < min_dist:
            min_dist = dist
            category = prototype[-1]
    return category

# Compute Euclidean distance between prototype and unknown
# sqrt((p1-u1)^2 + (p2-u2)^2 + ... + (pn-un)^2)
def compute_dist(prototype, unknown):
    if len(prototype) != len(unknown) + 1:
        raise DataError('The number of data points in the prototype did not match the number of data points in the unknown')
    sumOfSquares = 0
    # Note: we must loop through the length of the unknown
    # array because the prototype array is one longer because
    # it also holds the category
    for i in range(len(unknown)):
        sumOfSquares += (float(prototype[i]) - float(unknown[i]))**2
    return math.sqrt(sumOfSquares)

if __name__ == "__main__":
    prototype_file = sys.argv[1]
    unknown_file = sys.argv[2]

    prototypes = process_prototypes(prototype_file)
    find_nearest_neighbor(prototypes, unknown_file)

File: nn/nn.cpp
// Caroline Danzi
// Dr. Zmuda
// CSE 465
// Nearest Neighbor Problem
// Help with vectors: http://www.cplusplus.com/reference/vector/vector/

#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <vector>
#include <math.h> // for pow, sqrt
#include <limits.h> // For INT_MAX

// Convert a string line with comma-separated values
// to a double array of those values
// stod: http://www.cplusplus.com/reference/string/stod/
std::vector<double> convertStrToDoubleVector(std::string line)
{
	std::vector<double> dataVector;
	int commaIndex = line.find_first_of(',');
	double datum;
	
	// Loop through as long as the line has 
	while(commaIndex != -1)
	{
		// The substring is the first number we need
		// Convert to double and push_back on vector
		datum = std::stod(line.substr(0, commaIndex));
		dataVector.push_back(datum);
		// Chop off the first part of string up to the comma
		line = line.substr(commaIndex + 1, line.length() - (commaIndex + 1));
		// Find the index of first comma, this is also the length of the substring
		commaIndex = line.find_first_of(',');
	}
	// Loop did not include the last number
	datum = std::stod(line);
	dataVector.push_back(datum);
	return dataVector;
}

// Input: the file name of the prototype file
// Output: A 2D array of the prototypes
// Help reading lines from files from: 
// http://stackoverflow.com/questions/7868936/read-file-line-by-line
// http://www.cplusplus.com/forum/windows/48212/
std::vector<std::vector<double>> parsePrototypes(char* fileName)
{
	// Open file for reading
	std::ifstream prototypeFile;
	prototypeFile.open(fileName);
	
	// This will be a vector of arrays
	std::vector<std::vector<double>> prototypes;
	
	// Loop through each line in the file, split it by ',' and add it to 
	// the vector of prototypes
	std::string line = "";
	while(getline(prototypeFile, line))
	{
		prototypes.push_back(convertStrToDoubleVector(line));
	}
	prototypeFile.close();
	return prototypes;
}

// Input: 1D array of prototype data and 1D array of unknown data
// Output: double representing Euclidean distance between the prototype and 
// the unknown: sqrt((p1-u1)^2 + (p2-u2)^2 + ... + (pn-un)^2)
double computeDistance(std::vector<double> prototype, std::vector<double> unknown)
{
	double sum = 0;
	for(int i = 0; i < unknown.size(); i++)
	{
		sum += pow((prototype[i] - unknown[i]), 2);
	}
	return sqrt(sum);
}

// Input: 2D array of prototypes, 1D array of unknown data
// Output: double representing category for this unknown
// The category is determined by the closest prototype to the unknown
double categorize(std::vector<std::vector<double>> prototypes, std::vector<double> unknown)
{
	double minDist = INT_MAX;
	double category;
	for(int i = 0; i < prototypes.size(); i++)
	{
		std::vector<double> prototype = prototypes[i];
		double dist = computeDistance(prototype, unknown);
		if(dist < minDist)
		{
			minDist = dist;
			category = prototype[prototype.size() - 1];
		}
	}
	return category;
}

// Input: the list of prototypes and the name of the unknown file
// Returns: nothing
// Prints the category for each unknown data set
void findNearestNeighbor(std::vector<std::vector<double>> prototypes, char* fileName)
{
	// Open file for reading
	std::ifstream unknownFile;
	unknownFile.open(fileName);
	std::string line = "";
	
	std::vector<double> unknown;
	double category;
	
	while(getline(unknownFile, line))
	{
		unknown = convertStrToDoubleVector(line);
		category = categorize(prototypes, unknown);
		std::cout << category << std::endl;
	}
}

int main(int argc, char* argv[])
{
	if(argc < 3)
	{
		std::cerr << "Expected two command line arguments" << std::endl;
	}
	// The first command line argument is the name of the file,
	// the second is the prototype file and the third is the unknown file 
	char* prototype_file = argv[1];
	char* unknown_file = argv[2];
	
	std::vector<std::vector<double>> prototypes = parsePrototypes(prototype_file);
	findNearestNeighbor(prototypes, unknown_file);
	return 0;
}